### PR TITLE
Check contiguity of iterators in make_array_view

### PR DESCRIPTION
**Pull request by masterleinad**
This is a slightly modified version of https://stackoverflow.com/questions/35004633/checking-if-a-sequence-container-is-contiguous-in-memory
to check if the provided range is contiguous.

### CODE REVIEW

File: include/deal.II/base/array_view.h [hunk 1]
```diff
@@ -618,6 +640,8 @@ make_array_view (ElementType *const begin, ElementType *const end)
 {
   Assert(begin <= end,
          ExcMessage("The beginning of the array view should be before the end."));
+  Assert(internal::ArrayViewHelper::is_contiguous(begin, end),
+         ExcMessage("The provided range isn't contiguous in memory!"));
```
Comment (bangerth) [inline include/deal.II/base/array_view.h]: I don't think you need to test here -- you give two pointers, of course the range is contiguous (by definition).

File: include/deal.II/base/array_view.h [hunk 2]
```diff
@@ -609,6 +630,7 @@ make_array_view (const Iterator begin, const Iterator end)
  *
  * @warning The pointers @p begin and @p end must bound (in the usual
  * half-open way) a contiguous in memory range of values.
+ * This assumption is checked in debug mode.
```
Comment (bangerth) [inline include/deal.II/base/array_view.h]: I think the existing text may be misleading -- the function makes the *assumption* that this is so, but there is no way to check this: you can pass any two pointers, the *user* has to make sure they're valid.

File: include/deal.II/base/array_view.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/base/array_view.h]: I agree.

File: include/deal.II/base/array_view.h (hunk 2 above)
Comment (masterleinad) [reply include/deal.II/base/array_view.h]: If we are just given `Iterators`, we are able to detect this. In particular, the test shows that a `std::deque` with few elements is fine while one with a few thousand elements is not.

Comment (bangerth) [reply include/deal.II/base/array_view.h]: Correct. What I meant to say (but didn't very well) is that you can check contiguity for general iterators. For pointers, however, contiguity is *always* guaranteed -- it's an inherent property of pointers. On the other hand, whether a pointer range points into the same object in memory is something that you cannot check -- it's a property that can only be determined if you know where these pointers come from.

**Comment by masterleinad** [comment]
I (hopefully) improved the phrasing of the additional comment.

**Comment by bangerth** [comment]
/run-tests
OK to merge once the tester is happy.

**Pull request merged**